### ISSUE TITLE
How to apply AMixer to 1D[B,C,H]?

### ISSUE THREAD

**Issue by kuaileyuandi** [raoyongming/AMixer#1]
Hello, Thank you very much for conceiving the AMixer model. Not only the amount of parameters is very small, but also a high accuracy rate is obtained, which is very innovative. Now AMixer is at two-dimensional level [B,C,H,W], but when the data is one-dimensional data [B,C,H], how to use AMixer model? I would like to ask a question: Can you provide an AMixer with a 1D MLP level for reference?

**Comment by raoyongming** [comment]
Hi, thanks for your interest in our work. We focus on image recognition tasks in this paper. I think maybe you can modify the weight bank to HxH if you want to apply AMixer to 1D tasks.

**Comment by kuaileyuandi** [comment]
> Hi, thanks for your interest in our work. We focus on image recognition tasks in this paper. I think maybe you can modify the weight bank to HxH if you want to apply AMixer to 1D tasks.
Thank you very much for your reply. The following is the 2D dimension AMixer provided by Dr. Rao, can you change it to 1D for reference?
class AdaSpatialMLP(nn.Module):
    def __init__(self, dim, n=196, k=16, r=4, num_heads=1, mode='softmax', post_proj=False, pre_proj=False, relative=False):
        super().__init__()

        self.relative = relative
        if not relative:
            self.weight_bank = nn.Parameter(torch.randn(k, n, n, dtype=torch.float32) * 0.02)
        else:
            h = w = int(math.sqrt(n))
            assert h * w == n
            # define a parameter table of relative position bias
            self.weight_bank = nn.Parameter(torch.randn(k, (2 * h - 1) * (2 * w - 1), dtype=torch.float32) * 0.02)  # 2*Wh-1 * 2*Ww-1, nH

            # get pair-wise relative position index for each token inside the window
            coords_h = torch.arange(h)
            coords_w = torch.arange(w)
            coords = torch.stack(torch.meshgrid([coords_h, coords_w]))  # 2, Wh, Ww
            coords_flatten = torch.flatten(coords, 1)  # 2, Wh*Ww
            relative_coords = coords_flatten[:, :, None] - coords_flatten[:, None, :]  # 2, Wh*Ww, Wh*Ww
            relative_coords = relative_coords.permute(1, 2, 0).contiguous()  # Wh*Ww, Wh*Ww, 2
            relative_coords[:, :, 0] += h - 1  # shift to start from 0
            relative_coords[:, :, 1] += w - 1
            relative_coords[:, :, 0] *= 2 * w - 1
            relative_position_index = relative_coords.sum(-1)  # Wh*Ww, Wh*Ww
            self.register_buffer("relative_position_index", relative_position_index)

        self.adapter = nn.Sequential(
            nn.Linear(dim, dim//r),
            nn.GELU(),
            nn.Linear(dim//r, k * num_heads)
        )

        self.k = k
        self.dim = dim
        self.num_heads = num_heads
        self.n = n
        self.mode = mode


        if pre_proj:
            self.pre_proj = nn.Linear(dim, dim)
        else:
            self.pre_proj = None

        if post_proj:
            self.post_proj = nn.Linear(dim, dim)
        else:
            self.post_proj = None

        print('[AdaSpatialMLP layer] k=%d, num_heads=%d, mode=%s, pos/pre-proj=%s/%s, relative=%s' % (k, self.num_heads, mode, pre_proj, post_proj, relative))
        
        
    def forward(self, x, mask=None):
        B, n, C = x.shape
        mix_policy = self.adapter(x).reshape(B, n, self.k, self.num_heads)

        if not self.relative:
            weight_bank = self.weight_bank
        else:
            weight_bank = self.weight_bank[:, self.relative_position_index.view(-1)].view(self.k, n, n)  # k,Wh*Ww,Wh*Ww
        
        if self.mode == 'softmax':
            mix_policy = torch.softmax(mix_policy, dim=2)
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
        elif self.mode == 'linear':
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
        elif self.mode == 'softmax-softmax':
            mix_policy = torch.softmax(mix_policy, dim=2)
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
            weight = torch.softmax(weight, dim=1)
        elif self.mode == 'linear-softmax':
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
            weight = torch.softmax(weight, dim=1)
        elif self.mode == 'linear-sigmoid':
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
            weight = torch.sigmoid(weight)
        elif self.mode == 'linear-normalize':
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
            weight = torch.nn.functional.normalize(weight, dim=1, p=2)
        elif self.mode == 'sigmoid':
            mix_policy = torch.sigmoid(mix_policy)
            weight = torch.einsum('bnkh,knm->bnmh', mix_policy, weight_bank)
        else:
            raise NotImplementedError

        if self.pre_proj is not None:
            x = self.pre_proj(x)
        
        x = x.reshape(B, n, self.num_heads, -1)
        x = torch.einsum('bnhc,bnmh->bmhc', x, weight).reshape(B,n,C)

        if self.post_proj is not None:
            x = self.post_proj(x)
        return x
